### DISCUSSION TITLE
How to call a generic method using meta.Method.ReturnType as generic parameter?

### DISCUSSION THREAD

**Discussion by MEmanuelsson** [metalama/Metalama#235] [Q&A] [answered]
Is there a way to use meta.Method.ReturnType as a generic parameter in an aspect (OverrideMethodAspect)? 

Inspired by your caching sample but using IDistributedCache and serializing/deserializing cached data using System.Text.Json, how can I call the following method where T should be set at compile time using the known ReturnType of the function? I've tried to find examples doing this but without any success:

`var deserializedData = JsonSerializer.Deserialize<T>(cachedData);`

cachedData in this case is a byte array.

**Comment by gfraiteur** [comment, marked as answer]
This is possible using a generic auxiliary template. See https://doc.metalama.net/conceptual/aspects/templates/auxilliary-templates#dynamic-invocation-of-generic-templates.

However, in your case, you should instead use the non-generic overload of ` JsonSerializer.Deserialize`, this is simpler.

You may also like our ready-made caching aspect. See https://doc.metalama.net/patterns/caching.

**Reply by MEmanuelsson** [reply]
I'll have a look at the generic auxiliary template 👍

My current solution is looking like this, using a non-generic overload of `JsonSerializer.Deserialize` that takes a `Stream ` as input, which I would like to avoid having to instantiate, since I guess this is not the optimal way to do this(?):

```
using var stream = new MemoryStream(cachedData);
var returnType = meta.Target.Method.ReturnType.ToType();
JsonSerializer.Deserialize(stream, returnType)
```

Just now also noticed I can pass a `ReadOnlySpan<byte>` in another overload, I guess this might be a better solution than the `Stream` solution(?):

```
var returnType = meta.Target.Method.ReturnType.ToType();
JsonSerializer.Deserialize(new ReadOnlySpan<byte>(cachedData), returnType)

```

Thanks again for quick response! Great library!
